### DISCUSSION TITLE
Enabling 2FA on 5.6.8 outputs a JS Page instead of activating it

### DISCUSSION THREAD

**Discussion by thearchivalone** [firefly-iii/firefly-iii#5449] [Firefly III] [answered]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [ ] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [ ] ... it's [definitely me, not you](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

serviceworker.js?v=5.6.8 opens in full plain text in both browser and PWA (on mobile) upon enabling 2FA. I don’t think this was intended functionality (at least I hope not). I can grab my logs when I get off of work if needed (mobile right now for several hours).

### Expected behaviour

2FA activates with no issues.

### Debug information

Debug information generated at 2021-12-28 12:38:24 America/Chicago for Firefly III version **5.6.8**.

| Scope | Version |
| --- | --- |
| Firefly III | 5.6.8 |
| Firefly III API | 1.5.5 |
| PHP | 8.0.12 |
| Host | Linux |

| System info | Value |
| --- | --- |
| Installation ID | 9b4d8d49-935f-448e-9c8c-11bbba4f70d0 |
| Using docker? | false |
| Layout | v1 |
| System TZ | America/Chicago |
| Browser TZ | America/Chicago |
| App environment | production |
| App debug mode | false |
| App cache driver | file |
| App logging | notice, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Interface | fpm-fcgi |
| Default language | en_US |
| Default locale | equal |
| BCscale | 24 |
| DB drivers | pgsql |
| Current driver | pgsql |
| DB version | 18 (exp. 18) |
| Login provider | eloquent  |
| Trusted proxies (.env) | ** |

| User info | Value |
| --- | --- |
| Session start | 2021-12-01 00:00:00 |
| Session end | 2021-12-31 23:59:59 |
| Session first | 2021-06-22 00:00:00 |
| User ID | 1 |
| User language | en_US |
| User locale | en_US |
| Attempt at "en_US.utf8" | 'en_US.utf8' |
| Attempt at "en_US.UTF-8" | 'en_US.UTF-8' |
| User agent | Mozilla/5.0 (iPhone; CPU iPhone OS 15_1_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.1 Mobile/15E148 Safari/604.1 |

### Steps to reproduce

- Log into Firefly
- Click Options from Hamburger Buns on Left
- Click Profile
- Click 2FA
- Go through the steps
- Get the serviceworker.js file to show

### Additional info

- Server: Caddy-V2
- PHP-FPM 8.0 Unix Socket through Reverse Proxy
- 2FA Pass (Password Store iOS)

**Comment by JC5** [comment]
I can't replicate this. It works fine.

**Reply by thearchivalone** [reply]
![52B9B756-4145-4386-A892-CF0C3489F182](https://user-images.githubusercontent.com/59133880/147601299-ef14827d-03ae-4dc8-abe8-ca4208f423a6.jpeg)
![11B265B4-1E77-4301-A635-6DD00F032E1A](https://user-images.githubusercontent.com/59133880/147601310-e67b8873-e0b7-485a-80aa-e34238c61e82.png)

Weirdest thing ever. Thanks for responding. Any suggestions on how to possibly fix it?

**Reply by JC5** [reply]
Not really. I tried on my phone as well. It works fine. Not sure yet, be sure to try this on your desktop as well.

**Reply by thearchivalone** [reply]
Like I said, once I’m off work, I can check the logs. I also plan on making some changes to how i’m serving my instance. My guess: a permission error.

**Reply by thearchivalone** [reply]
I just had a thought: I moved my server hosts recently (upgraded to bigger at a cheaper rate) and exported my firefly iii db content from one to the other. The original one had 2FA enabled but on transfer and upgrade, that ended up being disabled. Is it possible that there could have been some  db corruption there causing the issue?

**Reply by JC5** [reply]
Sure, but that's nearly impossible.

Having 2FA is handled by a value in the users table. For such a value to drop is pretty unique to begin with. But the serviceworker.js showing up is something that's code-based, handled [here](https://github.com/firefly-iii/firefly-iii/blob/main/app/Support/Steam.php#L84).

So you getting redirected to the service worker script is something else, and I'm not sure yet what it could be.

**Reply by thearchivalone** [reply]
I'm home and immediately grabbed some food and looked at my logs. SELinux didn't catch any denials from it and the Firefly-iii logs didn't show anything out of the ordinary other than the normal audits and my transactions throughout the day. I checked my server log and found that there was a 304 Not Modified Error for serviceworker.js that had occurred 15 times over the last three days but this was the first time it actually affected my usage directly. About to test on my PC.

`[28/Dec/2021:21:10:20 -0600] "GET /serviceworker.js?v=5.6.8 HTTP/2.0" 304 0`

**Reply by thearchivalone** [reply]
So, with 2FA enabled, I'm getting something interesting. On my PC, I can enable and log in fine with it but on mobile, it's telling me that the code is wrong despite both mobile and the PC having the same 2FA methods with the same generated code.

Here's something even more interesting: It just worked on mobile but required the code to change from the first number for it to accept it. The first generated code is not taking but the second one seems fine. Hmmmmm

**Reply by JC5** [reply]
That's by design. You can't reuse codes twice to prevent replay attacks.

**Reply by thearchivalone** [reply]
I like that for the most part. Thanks for being patient and bearing with me here. Overall, I love my current password manager but this is proving that it's definitely got some limitations. It not syncing properly or well with your app along with a couple other ones just means I'll possibly be looking at switching again for security and sanity's sake.

Thanks again. ❤️

**Reply by JC5** [reply]
No problem! Let me know if I can do anything. Personally a big fan of 1Password, if it's any help.

Glad to hear you like Firefly's III! 🥰

**Comment by thearchivalone** [comment, marked as answer]
For anyone else running into this weird issue, the serviceworker showing up issue looks like it's mobile only and if there's at least 5 seconds or so left on your 2FA, just wait to input it. Also, a suggestion made by another self-hoster is to check your server's time. Mine's in sync with my localtime but Cloudflare, my DNS, is adding a few seconds of delay as a security measure for any web app running through it.
